### ISSUE TITLE
[Feature] Support deterministic inference with Batch Invariant Ops

### ISSUE THREAD

**Issue by Fridge003** [sgl-project/sglang#10278]
### Checklist

- [ ] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [ ] 2. Please use English, otherwise it will be closed.

### Motivation

-[ ] Support torch flex attention backend #9947
-[ ] Enable batch invariant mode

Reference:
https://thinkingmachines.ai/blog/defeating-nondeterminism-in-llm-inference/
https://github.com/thinking-machines-lab/batch_invariant_ops

**Comment by skyzh** [comment]
Hey, I'm interested in working on part of the issue but I have no idea where to start. Does "Support torch flex attention backend" mean that we need to wait for upstream support for deterministic kernels before we can integrate it with sglang? Would be great if you can have some pointers to the code we're going to patch in sglang. Thanks!

**Comment by Fridge003** [comment]
> Hey, I'm interested in working on part of the issue but I have no idea where to start. Does "Support torch flex attention backend" mean that we need to wait for upstream support for deterministic kernels before we can integrate it with sglang? Would be great if you can have some pointers to the code we're going to patch in sglang. Thanks!

Thanks for your interest~ We feel appreciated if you can help us

There are several kernels we need for deterministic inference:
RMSNorm, Matmul, Attention

For RMSNorm and Matmul kernels, we can directly apply the ones provided in https://github.com/thinking-machines-lab/batch_invariant_ops/tree/main/batch_invariant_ops, with a single operation `set_batch_invariant_mode(True)`.

But for deterministic attention kernels, we need to work on how to obtain them.  For flex attention we don't need to wait for upstream support since it's already supported. But for other kernels like flashinfer or triton, we need some extra development on how to make them deterministic.

Currently SGLang doesn't support flex attention as the attention backend. One way of achieving deterministic inference would be first supporting flex attention, and then apply some simple modifications similar to https://github.com/vllm-project/vllm/pull/24583. The main blocker for this approach is to integrate flex attention into sglang.

There has been a related PR #9947, which is a naive example of supporting flex attention in sglang. However it's not usable due to slow speed. To make flex attention usable, we might need some extra management on attention masks, so that the kernel is compiled once for each forward batch.

**Comment by skyzh** [comment]
did a quick test on flex attention: it seems like `enable_batch_invariant_mode` already yields a good enough deterministic results when flex attention is used. I had a branch with both https://github.com/sgl-project/sglang/pull/9947 and https://github.com/sgl-project/sglang/pull/10393 merged plus my change to enable batch invariant mode, and I ran the script with the qwen3-0.6b model. They seem to produce a stable 1 unique response. I have a small GPU so I reduced the batch size to [1, 10] but I don't think it matters.

https://github.com/skyzh/sglang/tree/skyzh/test-deterministic

```
python3 -m sglang.launch_server --model Qwen/Qwen3-0.6B --attention-backend flex_attention --mem-fraction-static 0.9
python3 -m sglang.test.test_deterministic --n-trials 10
```

It gives:

```
Total samples: 10, Unique samples: 1
```

I can follow up with submitting my patch once #9947 lands. I also have two questions:

- is there a central place where we can enable batch invariant ops? like vLLM's worker.py?
- do we want to make batch invariant ops configurable or just enable it globally for everyone?

Also, do we have plan to statically compile all the kernels in #9947? It looks like we dynamically generate the masks and kernels so the determinism is not guaranteed. I suppose having compiled kernels are required to ensure determinism.

Thanks!

**Comment by Fridge003** [comment]
Hi @skyzh, I opened this POC PR: #10417 two days ago, and many related details can be found in this PR.

- For batch invariant ops, I plan to put it at `model_runner.py`, which seems working well
- Batch invariant ops should be configurable. I will add a server argument for this later.
- For flex attention backend, we will need statically compiled kernels. But how to reach this is still under discussion. Currently we are trying to use flashinfer/triton kernels instead since flex attention will bring some complexity.
- In #10417 I added a harder test for determinism (--test-mode mixed). I strongly recommend running your code on the harder test and see whether the outputs are consistent (especially for the long prompt input)

Thanks!

**Comment by keyboardAnt** [comment]
@Fridge003, thanks for your awesome contribution! 👏

Are there plans to support non-greedy sampling for speculative decoding drafters (#9877)?

**Comment by Fridge003** [comment]
> [@Fridge003](https://github.com/Fridge003), thanks for your awesome contribution! 👏
> 
> Are there plans to support non-greedy sampling for speculative decoding drafters ([#9877](https://github.com/sgl-project/sglang/issues/9877))?

Do you mean supporting deterministic inference for spec decoding? We don't have that plan by now.

**Comment by keyboardAnt** [comment]
> > [@Fridge003](https://github.com/Fridge003), thanks for your awesome contribution! 👏
> > 
> > Are there plans to support non-greedy sampling for speculative decoding drafters ([#9877](https://github.com/sgl-project/sglang/issues/9877))?
> 
> Do you mean supporting deterministic inference for spec decoding? We don't have that plan by now.

Got it. Thanks so much for your open-source work regardless!

And yeah, my question is about making speculative decoding fully reproducible by applying the non-greedy sampling method [described](https://lmsys.org/blog/2025-09-22-sglang-deterministic/#reproducible-non-greedy-sampling) in your blogpost. This should yield deterministic acceptance and help address https://github.com/sgl-project/sglang/issues/8391#issuecomment-3206540788.

**Comment by Fridge003** [comment]
> > > [@Fridge003](https://github.com/Fridge003), thanks for your awesome contribution! 👏
> > > Are there plans to support non-greedy sampling for speculative decoding drafters ([#9877](https://github.com/sgl-project/sglang/issues/9877))?
> > 
> > 
> > Do you mean supporting deterministic inference for spec decoding? We don't have that plan by now.
> 
> Got it. Thanks so much for your open-source work regardless!
> 
> And yeah, my question is about making speculative decoding fully reproducible by applying the non-greedy sampling method [described](https://lmsys.org/blog/2025-09-22-sglang-deterministic/#reproducible-non-greedy-sampling) in your blogpost. This should yield deterministic acceptance and help address [#8391 (comment)](https://github.com/sgl-project/sglang/issues/8391#issuecomment-3206540788).

I just looked through #8391, seems the new sampling method is worth a try for reproducible speculative decoding.
I will add an item for this task.

**Comment by hebiao064** [comment]
For FA3, we can use fixed kv split size by feeding into num_splits = seq_len_k // 4096

@zminglei could u pls give a try?

**Comment by zminglei** [comment]
> For FA3, we can use fixed kv split size by feeding into num_splits = seq_len_k // 4096
> 
> [@zminglei](https://github.com/zminglei) could u pls give a try?

Sure, let me take a look

**Comment by linlang1837** [comment]
The autotune of Triton can lead to non-deterministic execution results. How does SGLang handle this?

**Comment by Fridge003** [comment]
> The autotune of Triton can lead to non-deterministic execution results. How does SGLang handle this?

Triton config should be fixed. It's natural that autotune will lead to different launching configs, thus different outputs.

**Comment by linlang1837** [comment]
What I mean is: if the configuration is not fixed, does SGLang do anything to handle Triton’s autotune to ensure that the same configuration is chosen every time? If not, does it mean that when a Triton kernel uses a non-fixed configuration, SGLang cannot guarantee deterministic results?

**Comment by Fridge003** [comment]
> What I mean is: if the configuration is not fixed, does SGLang do anything to handle Triton’s autotune to ensure that the same configuration is chosen every time? If not, does it mean that when a Triton kernel uses a non-fixed configuration, SGLang cannot guarantee deterministic results?

SGLang will avoid using autotune. For triton kernels, we will fix configurations like block sizes or num warps for the same model, so deterministic results can be obtained.

**Comment by hebiao064** [comment]
Smal task: Use FA3 when deterministic enabled and no attention backend being set @zminglei could you please help with it

cc @Fridge003

**Comment by Edenzzzz** [comment]
After this PR https://github.com/sgl-project/sglang/pull/10721, flashinfer seems to support determinism with radix cache?

**Comment by Fridge003** [comment]
@Edenzzzz We detected some indeterminism after merging this PR, so it was temporarily reverted #11728. The idea of this PR is correct, so it can be merged again after fixing bugs.

**Comment by hlu1** [comment]
A task related to Deepseek V3.2: add an option to use the pytorch topk for deterministic topk results. Once a deterministic fused kernel is available from Nvidia, we can use integrate with that one too. The current custom topk kernel, both unfused and fused, is non-deterministic which results in variance in accuracy benchmarks.

**Comment by TheDarkNight21** [comment]
Hello I hope all is well. I am looking to help in any way shape or form on this project but do not know where to start. Are there any pending issues I could take a look at? I am mainly interested in trying to implement a more efficient batch invariant kernel but if there are other pressing manners I am more than willing to help.
